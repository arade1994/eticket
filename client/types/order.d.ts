import { type Ticket } from "./ticket";

export interface Order {
  id: string;
  userId: string;
  status: OrderStatus;
  expiresAt: Date;
  version: number;
  ticket: Ticket;
}

export declare enum OrderStatus {
  Created = "created",
  Cancelled = "cancelled",
  AwaitingPayment = "awaiting:payment",
  Complete = "complete",
}
