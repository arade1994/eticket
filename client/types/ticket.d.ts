export interface Ticket {
  id: string;
  title: string;
  price: number;
  userId: string;
  category: string;
  version: number;
}
