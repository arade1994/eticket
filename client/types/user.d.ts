import { type IncomingMessage } from "http";

export interface User {
  id: string;
  firstName: string;
  lastName: string;
  age: number;
  email: string;
  password: string;
}

export interface Rating {
  id: string;
  comment: string;
  rate: number;
  userId: string;
  ratedUserId: string;
}

export interface RequestWithUser extends IncomingMessage {
  currentUser?: { id: string; email: string };
}
