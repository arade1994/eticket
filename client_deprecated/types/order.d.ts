import { type Ticket } from "./ticket";

export interface Order {
  id: string;
  userId: string;
  status: OrderStatus;
  expiresAt: string;
  version: number;
  ticket: Ticket;
}

export declare enum OrderStatus {
  Created = "created",
  Cancelled = "cancelled",
  AwaitingPayment = "awaiting:payment",
  Complete = "complete",
}
