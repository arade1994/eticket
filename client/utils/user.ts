import { Ticket } from "../types/ticket";
import { User } from "../types/user";

export const getNumOfCreatedTickets = (user: User, tickets: Ticket[]) =>
  tickets?.filter((ticket) => ticket.userId === user.id)?.length;
