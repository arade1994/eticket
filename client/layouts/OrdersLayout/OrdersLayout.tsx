import { ChangeEvent, useCallback, useMemo, useState } from "react";
import OrdersFilters from "../../components/Orders/OrdersFilters/OrdersFilters";
import { SingleValue } from "react-select";
import { filterOrders } from "../../utils/orders";
import { Order } from "../../types/order";
import { User } from "../../types/user";
import classes from "./OrdersLayout.module.scss";
import OrdersTable from "../../components/Orders/OrdersTable/OrdersTable";

interface Props {
  orders: Order[];
  users: User[];
}

const OrdersLayout: React.FC<React.PropsWithChildren<Props>> = ({
  orders,
  users,
}) => {
  const [searchText, setSearchText] = useState("");
  const [selectedUser, setSelectedUser] = useState({
    value: "-",
    label: "Any user",
  });
  const [selectedStatus, setSelectedStatus] = useState({
    value: "-",
    label: "Any status",
  });

  const handleChangeSearchText = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => setSearchText(event.target.value),
    []
  );

  const handleSelectUser = useCallback(
    (
      option: SingleValue<{
        value: string;
        label: string;
      }>
    ) =>
      setSelectedUser({
        value: option?.value ?? "-",
        label: option?.label ?? "Any user",
      }),
    []
  );

  const handleSelectStatus = useCallback(
    (
      option: SingleValue<{
        value: string;
        label: string;
      }>
    ) =>
      setSelectedStatus({
        value: option?.value ?? "-",
        label: option?.label ?? "Any status",
      }),
    []
  );

  const handleResetFilters = useCallback(() => {
    setSearchText("");
    setSelectedUser({ value: "-", label: "Any user" });
    setSelectedStatus({ value: "-", label: "Any status" });
  }, []);

  const filteredOrders = useMemo(
    () =>
      filterOrders(
        orders,
        searchText,
        selectedUser.value,
        selectedStatus.value
      ),
    [orders, searchText, selectedUser.value, selectedStatus.value]
  );

  const userOptions = useMemo(
    () =>
      users.map((user) => ({
        value: user.id,
        label: `${user.firstName} ${user.lastName}`,
      })),
    [users]
  );

  return (
    <div className={classes.ordersContent}>
      <OrdersFilters
        searchText={searchText}
        selectedUser={selectedUser}
        selectedStatus={selectedStatus}
        userOptions={userOptions}
        onChangeSearchText={handleChangeSearchText}
        onSelectUser={handleSelectUser}
        onSelectStatus={handleSelectStatus}
        onResetFilters={handleResetFilters}
      />
      {filteredOrders.length ? (
        <OrdersTable users={users} orders={orders} />
      ) : (
        <h1>No orders yet! Please go to tickets page and order one.</h1>
      )}
    </div>
  );
};

export default OrdersLayout;
