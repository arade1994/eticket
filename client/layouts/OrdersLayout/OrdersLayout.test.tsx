import { render } from "@testing-library/react";
import { describe, expect, test } from "vitest";
import OrdersLayout from "./OrdersLayout";
import mockDb from "../../mock/db.json";
import { Order } from "../../types/order";
import { User } from "../../types/user";

const mockOrders = mockDb.orders as unknown as Order[];
const mockUsers = mockDb.users as unknown as User[];

describe("<OrdersLayout />", () => {
  test("it should match a snapshot when rendered by default", () => {
    const { baseElement } = render(
      <OrdersLayout orders={mockOrders} users={mockUsers} />
    );

    expect(baseElement).toMatchSnapshot();
  });
});
