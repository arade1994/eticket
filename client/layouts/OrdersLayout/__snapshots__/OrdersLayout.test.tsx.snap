// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<OrdersLayout /> > it should match a snapshot when rendered by default 1`] = `
<body>
  <div>
    <div
      class="_ordersContent_72370e"
    >
      <div
        class="_ordersFilters_9576c6"
      >
        <input
          id="searchInput"
          placeholder="Insert title"
          type="text"
          value=""
        />
        <div
          class="css-b62m3t-container"
          id="userSelect"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-2-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
            role="log"
          />
          <div
            class="css-13cymwt-control"
          >
            <div
              class="css-hlgwow"
            >
              <div
                class="css-1dimb5e-singleValue"
              >
                Any user
              </div>
              <div
                class="css-19bb58m"
                data-value=""
              >
                <input
                  aria-activedescendant=""
                  aria-autocomplete="list"
                  aria-expanded="false"
                  aria-haspopup="true"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class=""
                  id="react-select-2-input"
                  role="combobox"
                  spellcheck="false"
                  style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="css-1wy0on6"
            >
              <span
                class="css-1u9des2-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class="css-1xc3v61-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <div
          class="css-b62m3t-container"
          id="statusSelect"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-3-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
            role="log"
          />
          <div
            class="css-13cymwt-control"
          >
            <div
              class="css-hlgwow"
            >
              <div
                class="css-1dimb5e-singleValue"
              >
                Any status
              </div>
              <div
                class="css-19bb58m"
                data-value=""
              >
                <input
                  aria-activedescendant=""
                  aria-autocomplete="list"
                  aria-expanded="false"
                  aria-haspopup="true"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class=""
                  id="react-select-3-input"
                  role="combobox"
                  spellcheck="false"
                  style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="css-1wy0on6"
            >
              <span
                class="css-1u9des2-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class="css-1xc3v61-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
      <table
        class="_ordersTable_eb8f56"
      >
        <thead
          class="_ordersTableHeader_eb8f56"
        >
          <th
            class="_headerData_eb8f56"
          >
            Ticket title
          </th>
          <th
            class="_headerData_eb8f56"
          >
            Price($)
          </th>
          <th
            class="_headerData_eb8f56"
          >
            User
          </th>
          <th
            class="_headerData_eb8f56"
          >
            Status
          </th>
          <th
            class="_headerData_eb8f56"
          >
            Expires at
          </th>
          <th
            class="_headerData_eb8f56"
          >
            Actions
          </th>
        </thead>
        <tbody>
          <tr
            class="_ordersTableRow_eb8f56"
          >
            <td
              class="_tableItem_eb8f56"
            >
              Football Match
            </td>
            <td
              class="_tableItem_eb8f56"
            >
              230.00
            </td>
            <td
              class="_tableItem_eb8f56"
            >
              Ante Rade
            </td>
            <td
              class="_tableItem_eb8f56"
            >
              created
            </td>
            <td
              class="_tableItem_eb8f56"
            >
              19:37
            </td>
            <td
              class="_tableItem_eb8f56"
            >
              <a
                class="_openOrderLink_eb8f56"
                href="/orders/5ccb"
              >
                Open
              </a>
            </td>
          </tr>
          <tr
            class="_ordersTableRow_eb8f56"
          >
            <td
              class="_tableItem_eb8f56"
            >
              Bethoveen Concerto
            </td>
            <td
              class="_tableItem_eb8f56"
            >
              540.00
            </td>
            <td
              class="_tableItem_eb8f56"
            >
              Ante Rade
            </td>
            <td
              class="_tableItem_eb8f56"
            >
              created
            </td>
            <td
              class="_tableItem_eb8f56"
            >
              19:37
            </td>
            <td
              class="_tableItem_eb8f56"
            >
              <a
                class="_openOrderLink_eb8f56"
                href="/orders/6253"
              >
                Open
              </a>
            </td>
          </tr>
          <tr
            class="_ordersTableRow_eb8f56"
          >
            <td
              class="_tableItem_eb8f56"
            >
              Concert
            </td>
            <td
              class="_tableItem_eb8f56"
            >
              120.00
            </td>
            <td
              class="_tableItem_eb8f56"
            >
              Ante Rade
            </td>
            <td
              class="_tableItem_eb8f56"
            >
              created
            </td>
            <td
              class="_tableItem_eb8f56"
            >
              19:37
            </td>
            <td
              class="_tableItem_eb8f56"
            >
              <a
                class="_openOrderLink_eb8f56"
                href="/orders/5c7c"
              >
                Open
              </a>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</body>
`;
